### ISSUE TITLE
Schemaspy vs Azure SQL Database

### ISSUE THREAD

**Issue by pbuxtonccuk** [schemaspy/schemaspy#640]
Apparently, when I point Schemaspy to an Azure SQL Database (usin the mssql05, or 08 dbtype) , syscomments doesn't exist in Azure SQL.   Is there a way I can prevent Schemaspy trying to extract these??

There doesn't seem an obvious way!

**Comment by npetzall** [comment]
I need to check, but might be as simple as creating a custom databaseType and override the selectCommets sqls. However I'm unsure if you get a fatal error or if it's just errors in the log.

There isn't an option to exclude comments?

**Comment by pbuxtonccuk** [comment]
I couldn't find an option to ignore the comments, and I tried the override.  I couldn't seem to get away from the error.   I can't believe I'm not the first to run into this.

**Comment by npetzall** [comment]
According to code you should get warnings and not errors, I don't have an azure db to test with so I can't confirm.

Could you supply the errors you are seeing?

**Comment by npetzall** [comment]
As for an override I think this should work.
Haven't tried it since I'm not close to a computer. If it doesn't work, we should probably make it work. 
```
extends=mssql08
selectCatalogsSql= 
selectSchemasSql=
selectTableCommentsSql=
selectColumnCommentsSql=
```

As an update to previous comment, you probably do get two error logs, they should be changed to warning and it relates to catalog comment and schema comment.

**Comment by npetzall** [comment]
Here's another override that might work. 
```
extends=mssql08
selectCatalogsSql=SELECT 'nothing' where 1=0
selectSchemasSql=SELECT 'nothing' where 1=0
selectTableCommentsSql=SELECT 'nothing' where 1=0
selectColumnCommentsSql=SELECT 'nothing' where 1=0
```

**Comment by pbuxtonccuk** [comment]
Unfortunately, having tried both, neither seem to work

So,my command line is 

java  -jar c:\schemaspy\schemaspy-6.1.1-SNAPSHOT.jar  -t azuresql  -dp c:\schemaspy\mssql-jdbc-6.4.0.jre8.jar -host xxx  -port 1433 -db yyy -o c:\schemaspy\output\elastacloud  -all -vizjs -norows -u aaa -p bbb -debug


azuresql.properties has been set to both of your options above separrately and both tried

with -debug, the following is apparent...

DEBUG - DbSpecificOption name: 'hostOptionalPort' value: 'xxx:1433' description: 'null'
DEBUG - DbSpecificOption name: 'db' value: 'xxx' description: 'database name'
DEBUG - supportsSchemasInTableDefinitions: true
DEBUG - supportsCatalogsInTableDefinitions: true
DEBUG - Catalog not provided, queried jdbc driver and got 'xxx'
INFO  - Connected to Microsoft SQL Server - 12.00.2531
INFO  - Gathering schema details
Gathering schema details...DEBUG - Including view CHECK_CONSTRAINTS: matches inclusion pattern '.*'
DEBUG - Creating View xx.INFORMATION_SCHEMA.CHECK_CONSTRAINTS
DEBUG - select text AS view_definition from syscomments sc, sysobjects so where sc.id=so.id and so.name=? [CHECK_CONSTRAINTS]
ERROR - select text AS view_definition from syscomments sc, sysobjects so where sc.id=so.id and so.name=:table
ERROR - SqlException
com.microsoft.sqlserver.jdbc.SQLServerException: Catalog view 'syscomments' is not supported in this version.
        at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:262)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1632)
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:600)
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:522)
        at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7225)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3053)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:247)
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:222)
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeQuery(SQLServerPreparedStatement.java:444)
        at org.schemaspy.input.dbms.service.ViewService.gatherViewDefinition(ViewService.java:86)
        at org.schemaspy.input.dbms.service.ViewService.gatherViewsDetails(ViewService.java:68)
        at org.schemaspy.input.dbms.service.DatabaseService.initViews(DatabaseService.java:211)
        at org.schemaspy.input.dbms.service.DatabaseService.gatherSchemaDetails(DatabaseService.java:92)
        at org.schemaspy.SchemaAnalyzer.analyze(SchemaAnalyzer.java:246)
        at org.schemaspy.SchemaAnalyzer.analyzeMultipleSchemas(SchemaAnalyzer.java:176)
        at org.schemaspy.SchemaAnalyzer.analyze(SchemaAnalyzer.java:118)
        at org.schemaspy.cli.SchemaSpyRunner.runAnalyzer(SchemaSpyRunner.java:98)
        at org.schemaspy.cli.SchemaSpyRunner.run(SchemaSpyRunner.java:87)
        at org.schemaspy.Main.main(Main.java:55)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)

**Comment by npetzall** [comment]
Ahhh, we seem to have used this in more places, so now you have issues with views.

I'll see if I can get some time tonight and create a azure configuration. I'm gonna assume I can sign up for a 7day trial or something. But I'll start with azure sql documentation.

**Comment by pbuxtonccuk** [comment]
Thanks!!  

It's also worth noting that MS have said they will deprecate syscomments anyway, maybe something to consider

https://docs.microsoft.com/en-us/sql/relational-databases/system-compatibility-views/sys-syscomments-transact-sql?view=sql-server-ver15

**Comment by npetzall** [comment]
I need to appolagise, I made a faulty assumption in the beginning that syscomments was comments and not views, functions, constraints, triggers and such.

I had a quick look but don't have the time to fix it tonight sadly. But since the replacement is available in sqlserver I'll start working on a mssql19 databaseType that should be azure compatible.

**Comment by npetzall** [comment]
Finally got some hours to fiddle with this.

I haven't finished testing it all yet. But syscomment have been replaced, also sys.extendedproperties.

But here it goes
```
extends=mssql08

selectViewSql=select sm.definition as view_definition from sys.objects so left join sys.sql_modules sm on so.object_id = sm.object_id where so.[type] = 'V' and so.name =:table
selectCheckConstraintsSql=select OBJECT_NAME(parent_object_id) as table_name, name AS constraint_name, definition as text from sys.check_constraints

selectCatalogsSql=SELECT cast(value as nvarchar(max)) as catalog_comment FROM fn_listextendedproperty(N'MS_Description', default, default, default, default, default, default)
selectSchemasSql=SELECT cast(value as nvarchar(max)) as schema_comment FROM fn_listextendedproperty(N'MS_Description', N'schema', :schema, default, default, default, default)
selectTableCommentsSql=SELECT objname as table_name, cast(value as nvarchar(max)) as comments FROM fn_listextendedproperty(N'MS_Description', N'schema', :schema, 'table', null, default, default)
selectColumnCommentsSql=select x.table_name, x.column_name, x.comments from ( select t.name as table_name, c.name as column_name, cast((SELECT value from fn_listextendedproperty(N'MS_Description', N'schema', SCHEMA_NAME(t.schema_id), 'table', t.name, 'column', c.name)) as NVARCHAR(MAX)) as comments from sys.tables t left join sys.columns c on (t.object_id = c.object_id) where SCHEMA_NAME(t.schema_id) = :schema ) x where not x.comments is null
selectViewCommentsSql=SELECT objname as view_name, cast(value as nvarchar(max)) as comments FROM fn_listextendedproperty(N'MS_Description', N'schema', :schema, 'view', null, default, default)
selectViewColumnCommentsSql=select x.view_name, x.column_name, x.comments from ( select v.name as view_name, c.name as column_name, cast((SELECT value from fn_listextendedproperty(N'MS_Description', N'schema', SCHEMA_NAME(v.schema_id), 'view', v.name, 'column', c.name)) as NVARCHAR(MAX)) as comments from sys.views v left join sys.columns c on (v.object_id = c.object_id) where SCHEMA_NAME(v.schema_id) = :schema ) x where not x.comments is null

selectRoutinesSql=SELECT i_s.routine_name, i_s.routine_type, i_s.data_type AS dtd_identifier, i_s.routine_body, i_s.routine_definition, i_s.is_deterministic, i_s.sql_data_access, NULL AS security_type, NULL AS sql_mode, (select cast(value as nvarchar(max)) from fn_listextendedproperty(N'MS_Description', N'schema', i_s.routine_schema, 'procedure', i_s.ROUTINE_NAME, default, default)) AS routine_comment FROM information_schema.routines i_s WHERE routine_schema = :schema
selectRoutineParametersSql=SELECT specific_name, parameter_name, data_type AS dtd_identifier, parameter_mode FROM information_schema.parameters WHERE specific_schema = :schema AND ordinal_position != 0 ORDER BY ordinal_position
```

**Comment by dj-xatra** [comment]
I'm trying Schemaspy with a SQL Server 2016 database. I also ran into the errors about comments. Eventually I found [this old Schemaspy bug report](https://sourceforge.net/p/schemaspy/bugs/146/). The properties suggested there made it work.

**Comment by npetzall** [comment]
@dj-xatra there are multiple database types for mssql.

mssql (oldest)
mssql05 (for 2005 and newer)
mssql08 (for 2008 and newer)

And the one above which is new and for Azure/2019 however not yet in schemaspy.

Since you are using 2016 and we actually run our test suite against 2017 with mssql08. It would probably work with `-t mssql08`

**Comment by dj-xatra** [comment]
I need to correct myself: the database version here is 17.9.1.

I did some more testing with different dbtypes. These are the warnings I get:

With mssql08:

WARN  - Failed to retrieve table comments using SQL 'SELECT i_s.TABLE_NAME, CAST(s.value AS NVARCHAR(MAX)) AS comments FROM INFORMATION_SCHEMA.Tables i_s INNER JOIN sys.extended_properties s ON s.major_id = OBJECT_ID(i_s.table_catalog + '.' + i_s.table_schema + '.' + i_s.table_name) AND s.class = 1 AND s.name = 'MS_Description' AND s.minor_id = 0 WHERE i_s.table_schema = :schema'
Invalid object name 'INFORMATION_SCHEMA.Tables'.
WARN  - Failed to retrieve stored procedure/function details using sql 'SELECT i_s.routine_name, i_s.routine_type, i_s.data_type AS dtd_identifier, i_s.routine_body, i_s.routine_definition, i_s.is_deterministic, i_s.sql_data_access, NULL AS security_type, NULL AS sql_mode, CAST(s.value AS NVARCHAR(MAX)) AS routine_comment FROM information_schema.routines i_s LEFT JOIN sys.extended_properties s ON s.major_id = OBJECT_ID(i_s.routine_catalog + '.' + i_s.routine_schema + '.' + i_s.routine_name) AND s.class = 1 AND s.name = 'MS_Description' AND s.minor_id = 0 WHERE routine_schema = :schema'
Invalid object name 'information_schema.routines'.
WARN  - Failed to retrieve stored procedure/function details using SQL 'SELECT specific_name, parameter_name, data_type AS dtd_identifier, parameter_mode FROM information_schema.parameters WHERE specific_schema = :schema AND ordinal_position != 0 ORDER BY ordinal_position'
Invalid object name 'information_schema.parameters'.

With extends mssql08 and tablecomment select from old bug report in properties:

WARN  - Failed to retrieve stored procedure/function details using sql 'SELECT i_s.routine_name, i_s.routine_type, i_s.data_type AS dtd_identifier, i_s.routine_body, i_s.routine_definition, i_s.is_deterministic, i_s.sql_data_access, NULL AS security_type, NULL AS sql_mode, CAST(s.value AS NVARCHAR(MAX)) AS routine_comment FROM information_schema.routines i_s LEFT JOIN sys.extended_properties s ON s.major_id = OBJECT_ID(i_s.routine_catalog + '.' + i_s.routine_schema + '.' + i_s.routine_name) AND s.class = 1 AND s.name = 'MS_Description' AND s.minor_id = 0 WHERE routine_schema = :schema'
Invalid object name 'information_schema.routines'.
WARN  - Failed to retrieve stored procedure/function details using SQL 'SELECT specific_name, parameter_name, data_type AS dtd_identifier, parameter_mode FROM information_schema.parameters WHERE specific_schema = :schema AND ordinal_position != 0 ORDER BY ordinal_position'
Invalid object name 'information_schema.parameters'.

With [mssql-jdbc-8.2.0](https://github.com/microsoft/mssql-jdbc): no warnings if I use this properties file.

extends=mssql
#https://docs.microsoft.com/en-us/sql/connect/jdbc/building-the-connection-url?view=sql-server-ver15
connectionSpec=jdbc:sqlserver://<hostOptionalPort>;databaseName=<db>

selectColumnCommentsSql=SELECT OBJECT_NAME(EXP.major_id) AS TABLE_NAME, C.name AS COLUMN_NAME, EXP.value AS comments \
FROM sys.extended_properties AS EXP \
LEFT OUTER JOIN sys.columns AS C \
ON C.object_id = EXP.major_id \
AND C.column_id = EXP.minor_id \
WHERE \
EXP.class_desc = 'OBJECT_OR_COLUMN' and \
C.name is not null

selectTableCommentsSql=SELECT OBJECT_NAME(EXP.major_id) AS TABLE_NAME, EXP.value AS comments \
FROM sys.extended_properties AS EXP \
LEFT OUTER JOIN sys.columns AS C \
ON C.object_id = EXP.major_id \
AND C.column_id = EXP.minor_id \
WHERE \
EXP.class_desc = 'OBJECT_OR_COLUMN' and \
EXP.minor_id = 0

**Comment by npetzall** [comment]
@dj-xatra that version number can't be correct. Version 15 is Sql server 2019. Which would make 17 Sql server 2021.

**Comment by npetzall** [comment]
Google and you may find. https://social.msdn.microsoft.com/Forums/sqlserver/en-US/895fcd84-6fe7-4a30-8deb-fca57c96995f/invalid-object-name-informationschematables?forum=transactsql

So the invalid object name. Might be a case sensitivity issue. So I'm gonna see how I can make it case sensitive and then correct the queries accordingly.

I quick check would be in sqlcmd or studio trying with both 
`select top(1) * from INFORMATION_SCHEMA.Tables` and  
`select top(1) * from INFORMATION_SCHEMA.TABLES`

From the docs it looks like INFORMATION_SCHEMA should be all caps.

**Comment by dj-xatra** [comment]
Sorry, checked version again. It's 13, so 2016.

I ran your checks. Collation here is Latin1_General_BIN. The Tables query doesn't work, the TABLES one does.

**Comment by npetzall** [comment]
@dj-xatra you can download a new snapshot in a cpl of minutes that should now work for you using `-t mssql08`

**Comment by dj-xatra** [comment]
Indeed, it works flawlessly. Thanks for your effort.

**Comment by npetzall** [comment]
@pbuxtonccuk I have limited testing abilities, but there is now a databaseType mssql17 which works with 2019 and and has avoided anything that isn't available in Azure SQL.

So if you could download and test with the latest snapshot and `-t mssql17` if you have any issue reopen this issue and I'll take another stab at it.

**Comment by pbuxtonccuk** [comment]
Hi Nils,

Thanks for that – I will try it and let you know as soon as I can.

Regards
Paul
